src/shell/mod.rs

```diff
@@ -1151,7 +1193,124 @@ impl Shell {
+                        }
+                    }
+                    if let Some(xwm) = self
+                        .xwayland_state
+                        .as_mut()
+                        .and_then(|state| state.xwm.as_mut())
+                    {
+                        {
+                            for window in set.workspaces[set.active]
+                                .tiling_layer
+                                .mapped()
+                                .map(|(w, _)| w)
+                                .chain(set.workspaces[set.active].floating_layer.space.elements())
+                            {
+                                if let Some(surf) = window.active_window().x11_surface() {
+                                    let _ = xwm.raise_window(surf);
+                                }
+                            }
+                            for window in set.sticky_layer.space.elements() {
+                                if let Some(surf) = window.active_window().x11_surface() {
+                                    let _ = xwm.raise_window(surf);
+                                }
+                            }
+                            if let Some(surf) = set.workspaces[set.active]
+                                .fullscreen
+                                .as_ref()
+                                .and_then(|f| f.surface.x11_surface())
+                            {
+                                let _ = xwm.raise_window(surf);
+                            }
+                        }
+                        for surface in &self.override_redirect_windows {
+                            let _ = xwm.raise_window(surface);
+                        }
+                    }
+
+                    let output_geo = output.geometry();
+                    Ok(Some(
+                        output_geo.loc
+                            + Point::from((output_geo.size.w / 2, output_geo.size.h / 2)),
+                    ))
+                } else {
+                    Ok(None)
+                }
+            }
+            WorkspaceMode::Global => {
+                for set in self.workspaces.sets.values_mut() {
+                    if let Some((_, workspace_delta)) = set.previously_active {
+                        match workspace_delta {
+                            WorkspaceDelta::Gesture(delta) => {
+                                if delta.abs() >= GESTURE_THRESHOLD {
+                                    set.activate(
+                                        set.active,
+                                        Some(WorkspaceDelta::Shortcut(Instant::now(), delta)),
+                                        &mut self.workspace_state.update(),
+                                    )?;
+                                } else {
+                                    set.activate_previous(
+                                        Some(WorkspaceDelta::Shortcut(Instant::now(), 1.0 - delta)),
```

Comment:
Clever trick to make the compositor run the rest of the animation itself. But this would result in the rest always taking 200ms (or whatever `ANIMATION_DURATION` ends up being) even if the remaining movement is just a tiny portion, right?

So we kinda have to either forge an `Instant` in the past (possible by substracting `Duration`) or add a `Duration` parameter to the `Shortcut` variant.

Also regarding my previous comment, I see now, why `activate` takes an `Option` to be able to insert these `Shortcut`-variants. So replacing this with a boolean wouldn't work, but I still would prefer getting rid of the `Option`.

Reply:
> Clever trick to make the compositor run the rest of the animation itself. But this would result in the rest always taking 200ms (or whatever ANIMATION_DURATION ends up being) even if the remaining movement is just a tiny portion, right?

Nope, Shortcut takes a f64 value for how far in the animation we should be starting from (from 0.0 to 1.0) and some math is done to make sure we're at the right place in the animation.

Reply:
> Also regarding my previous comment, I see now, why activate takes an Option to be able to insert these Shortcut-variants. So replacing this with a boolean wouldn't work, but I still would prefer getting rid of the Option.

Would maybe having WorkspaceDelta impl Default be a good compromise? Then you could just insert WorkspaceDelta::default or whatever instead of None

Reply:
> Would maybe having WorkspaceDelta impl Default be a good compromise? Then you could just insert WorkspaceDelta::default or whatever instead of None

I don't think default covers it well given the two variants.

And `WorkspaceDelta::shortcut()` is only marginally longer than `WorkspaceDelta::default()` and tells you much more.